pkg/credentialstores/common.go

```diff
@@ -0,0 +1,129 @@
+package credentialstores
+
+import (
+	"net/url"
+	"strings"
+
+	"github.com/openshift/image-registry/pkg/kubernetes-common/credentialprovider"
+
+	"github.com/golang/glog"
+)
+
+// MultiError is a wrap for multiple errors.
+type MultiError struct {
+	errors []string
+}
+
+// Append adds a new error.
+func (m *MultiError) Append(e error) {
+	m.errors = append(m.errors, e.Error())
+}
+
+// Error exists to comply with golang's error interface.
+func (m *MultiError) Error() string {
+	return strings.Join(m.errors, " ")
+}
+
+// basicCredentialsFromKeyring extract basicAuth information from provided
+// keyring. If keyring does not contain information for the provided URL, empty
+// strings are returned instead.
+func basicCredentialsFromKeyring(keyring credentialprovider.DockerKeyring, target *url.URL) (string, string) {
+	regURL := getURLForLookup(target)
+	if configs, found := keyring.Lookup(regURL); found {
+		glog.V(5).Infof(
+			"Found secret to match %s (%s): %s",
+			target, regURL, configs[0].ServerAddress,
+		)
+		return configs[0].Username, configs[0].Password
+	}
+
+	// do a special case check for docker.io to match historical lookups
+	// when we respond to a challenge
+	if regURL == "auth.docker.io/token" {
+		glog.V(5).Infof(
+			"Being asked for %s (%s), trying %s, legacy behavior",
+			target, regURL, "index.docker.io/v1",
+		)
+		return basicCredentialsFromKeyring(
+			keyring, &url.URL{Host: "index.docker.io", Path: "/v1"},
+		)
+	}
+
+	// docker 1.9 saves 'docker.io' in config in f23, see
+	// https://bugzilla.redhat.com/show_bug.cgi?id=1309739
+	if regURL == "index.docker.io" {
```

Comment:
This logic is inside keyring.Lookup

Reply:
This has relation with this "CredentialStore containing multiple credentials" topic we have been discussing in other comments.